common/src/MCWeight.cxx

```diff
@@ -143,12 +143,102 @@ bool MCScaleVariation::process(Event & event){
 
 
 
+
+MCMuonScaleFactor::MCMuonScaleFactor(uhh2::Context & ctx,
+                                     const std::string & sf_file_path,
+                                     const std::string & sf_name,
+                                     float sys_error_percantage,
+                                     const std::string & weight_postfix,
+                                     const std::string & muons_handle_name): 
+  h_muons_            (ctx.get_handle<std::vector<Muon>>(muons_handle_name)),
+  h_muon_weight_      (ctx.declare_event_output<float>("weight_sfmu_" + weight_postfix)),
+  h_muon_weight_up_   (ctx.declare_event_output<float>("weight_sfmu_" + weight_postfix + "_up")),
+  h_muon_weight_down_ (ctx.declare_event_output<float>("weight_sfmu_" + weight_postfix + "_down")),
+  sys_error_percantage_(sys_error_percantage/100.)
+{
+  auto dataset_type = ctx.get("dataset_type");
+  bool is_mc = dataset_type == "MC";
+  if (!is_mc) {
+    cout << "Warning: MCMuonScaleFactor will not have an effect on "
+         <<" this non-MC sample (dataset_type = '" + dataset_type + "')" << endl;
+    return;
+  }
+
+  TFile sf_file(sf_file_path.c_str());
+  if (sf_file.IsZombie()) {
+    throw runtime_error("Scale factor file for muons not found: " + sf_file_path);
+  }
+  
+  sf_hist_.reset((TH2*) sf_file.Get((sf_name + "/pt_abseta_ratio").c_str()));
+  if (!sf_hist_.get()) {
+    throw runtime_error("Scale factor directory not found in file: " + sf_name);
+  }
+  sf_hist_->SetDirectory(0);
+  eta_min_ = sf_hist_->GetYaxis()->GetXmin();
+  eta_max_ = sf_hist_->GetYaxis()->GetXmax();
+  pt_min_  = sf_hist_->GetXaxis()->GetXmin();
+  pt_max_  = sf_hist_->GetXaxis()->GetXmax();
+
+  sys_direction_ = 0;
+  auto ctx_mu_sys_ = ctx.get("MCMuonScaleFactorSys", "");
+  if (ctx_mu_sys_ == "up") {
+    sys_direction_ = 1;
+  } else if (ctx_mu_sys_ == "down") {
+    sys_direction_ = -1;
+  }
+}
+
+bool MCMuonScaleFactor::process(uhh2::Event & event) {
+
+  if (!sf_hist_) {  
+    event.set(h_muon_weight_,       1.);
+    event.set(h_muon_weight_up_,    1.);
+    event.set(h_muon_weight_down_,  1.);
+  }
+
+  const auto & muons = event.get(h_muons_);
+  float weight = 1., weight_up = 1., weight_down = 1.;
+  for (const auto & mu : muons) {
+    float eta = mu.eta();
+    float pt = mu.pt();
+    if (eta_min_ < eta &&
+        eta_max_ > eta &&
+        pt_min_  < pt &&
+        pt_max_  > pt) {
+      int bin   = sf_hist_->FindFixBin(pt, eta);
+      float w   = sf_hist_->GetBinContent(bin);
+      float err = sf_hist_->GetBinError(bin);
+
+      weight      *= w;
+      weight_up   *= (w*(1+sys_error_percantage_) + err);
+      weight_down *= (w*(1-sys_error_percantage_) - err);
```

Comment:
Are the formulas correct here? And is sys_error_percantage really given in percent? If you say, that sys_error_percantage=1 gives you 1% of the systematic uncertainty, I would put sys_error_percantage to 100 to apply the full uncertainty (according to your comment in the header file). But then, these formulas here do not make sense for me. I would expect something like:

weight_up   *= w*(1 + sys_error_percantage_/100. * err);
weight_down *= w*(1 - sys_error_percantage_/100. * err);